packages/@aws-cdk/core/lib/stack.ts

```diff
@@ -65,7 +67,23 @@ export class Stack extends Construct implements ITaggable {
    * @param construct The construct to start the search from.
    */
   public static of(construct: IConstruct): Stack {
-    return _lookup(construct);
+    // we want this to be as cheap as possible. cache this result by mutating
+    // the object. anecdotally, at the time of this writing, @aws-cdk/core unit
+    // tests hit this cache 1,112 times, @aws-cdk/aws-cloudformation unit tests
```

Comment:
But how much time does it save?

Reply:
Not sure, but since `Stack.of` is a public API and heavily used both by the framework and by users, I think it's worth the opportunistic optimization.